### ISSUE TITLE
enhance quarto syntax highlighting

### ISSUE THREAD

**Issue by milanglacier** [quarto-dev/quarto-nvim#25]
The syntax highlight provided from treesiter's markdown parser is far away from mature.

Should prefer the old-fashined pandoc-syntax clone (i.e quarto-vim).

**Comment by jmbuhr** [comment]
While I do also still have some issues with TS, namely indentation (which is experimental), we need it (and it's speed) for the extraction of code for code completion and other LSP features via https://github.com/jmbuhr/otter.nvim.

However, there is nothing stopping you from using one of the old regex-based highlighters on top (see last option in this code chunk: https://github.com/nvim-treesitter/nvim-treesitter#modules)

That being said, in the long run it would be great to get our own pandoc/quarto markdown TS parser to fix those issues. The emacs people will also be happy about this since they have now also adopted TS.

What are the concrete issues you are getting with the current parser?

**Comment by milanglacier** [comment]
Almost everything.

The only thing treesitter does better is the injected languages in the code block.

Several most significant part:

1. Math equation highlight.

Pandoc-syntax has pretty decent highlight, while treesitter highlight `0%` for the math equation.

<img width="1074" alt="Screen Shot 2023-01-23 at 02 34 24" src="https://user-images.githubusercontent.com/45728125/213987489-f210d0dc-e6e2-4ea2-b6af-ffd1fbbeb418.png">

<img width="598" alt="Screen Shot 2023-01-23 at 02 34 39" src="https://user-images.githubusercontent.com/45728125/213987500-f5887de2-5ee6-4668-97f2-f9bca3d6157f.png">

2. General Latex syntax highlight

<img width="281" alt="Screen Shot 2023-01-23 at 02 36 46" src="https://user-images.githubusercontent.com/45728125/213987708-11edd93b-59a8-4fba-8cc3-2c8db5a1257e.png">

3. General markup identifier highlight

Treesitter doesn't have proper highlight for **bold**, *italic*, `inline code`, and 

>quote.

<img width="290" alt="Screen Shot 2023-01-23 at 02 41 58" src="https://user-images.githubusercontent.com/45728125/213988066-faa7e940-40ee-406f-9df1-409aae88cca7.png">

<img width="294" alt="Screen Shot 2023-01-23 at 02 42 15" src="https://user-images.githubusercontent.com/45728125/213988069-601ed09a-0a8d-4cbf-8b2d-f64e660a270a.png">

**Comment by milanglacier** [comment]
Beside, instead of manually tweaking `quarto`'s filetype to `markdown`, it is better to patch to the nvim-treesitter uptream:

https://github.com/nvim-treesitter/nvim-treesitter/blob/6cf93175bf011ca899e494376838ae147870a1eb/lua/nvim-treesitter/parsers.lua#L17

such that `quarto` filetype automatically uses `markdown`'s treesitter syntax.

Setting `quarto`'s filetype to `markdown`  would lead to many misconfigurations.

**Comment by jmbuhr** [comment]
1. Math is something we could and should solve with a latex language injection like the markdown grammar already does for html blocks here: https://github.com/MDeiml/tree-sitter-markdown/blob/split_parser/tree-sitter-markdown/queries/injections.scm I am guessing the maintainer would be open to a PR in this case, even though it is a bit more than the commonmark spec. Or maybe this is already where we should fork a new pandoc/quarto TS grammar. What are your thoughts on this?
2. For inline LaTeX without some delimiter this won't work.
3. General highlighting looks fine on my end:
![image](https://user-images.githubusercontent.com/17450586/214000781-b78ee6cc-7a29-49e4-9e93-65aadfd772aa.png)
or with conceallevel=3:
![image](https://user-images.githubusercontent.com/17450586/214000849-716c55d9-943d-4613-a496-fa7efff2ad16.png)

You can reproduce this with the kickstarter configuration: https://github.com/jmbuhr/quarto-nvim-kickstarter
maybe TS syntax highlighting was not enabled?

The quarto filetype is quite a recent addition to neovim, but I think we can safely assume that the overlap between neovim and quarto users and users that like to get the latest version is very large. So now is the time to make use of the quarto filetype. Would you like to do the honers and make the PR to nvim-treesitter?

**Comment by jmbuhr** [comment]
Will merge https://github.com/quarto-dev/quarto-nvim/pull/26 once the TS PR is merged.

**Comment by jmbuhr** [comment]
Regarding your issue with rose-pine I don't understand why it is not working either. It has the correct highlighting group:
![image](https://user-images.githubusercontent.com/17450586/214016498-97f171b8-b8c1-4fe1-af84-56fd8f83b0bb.png)
and rose-pine has highlighting defined for them: https://github.com/rose-pine/neovim/blob/845a6ad5443e3559dde42910c4523a5835c9233b/lua/rose-pine/theme.lua#L287

**Comment by jmbuhr** [comment]
Small update: I also noticed very recently that formatting has gotten a lot better, but haven't quite figured out which update caused this.

**Comment by milanglacier** [comment]
> Regarding your issue with rose-pine I don't understand why it is not working either. It has the correct highlighting group:
> ![image](https://user-images.githubusercontent.com/17450586/214016498-97f171b8-b8c1-4fe1-af84-56fd8f83b0bb.png)
> and rose-pine has highlighting defined for them: https://github.com/rose-pine/neovim/blob/845a6ad5443e3559dde42910c4523a5835c9233b/lua/rose-pine/theme.lua#L287

You can replicate the issue right?

And I have tried with other popuplar themes, they have more or less the same problem. Cannot properly highlight even the basic markup like bold, italics (including tokyonight.nvim, catppuccin.nvim).

**Comment by jmbuhr** [comment]
With rose-pine I can, but the kickstarter configuration is using catppuccin and that works (see screenshots). If catppuccin doesn't work for you out of the box, please try it with the kickstarter configuration so that we can compare a common ground.

**Comment by jmbuhr** [comment]
That sounds like an nvimR error.

I honestly think the vim-slime way is better because there are less moving parts, but feel free to show your workflow! :) You can see my different workflows with it here: https://youtube.com/playlist?list=PLabWm-zCaD1axcMGvf7wFxJz8FZmyHSJ7

**Comment by jmbuhr** [comment]
And I'll always prefer the more general solution, both for elegance and because quarto is meant to work with different languages, not just R.

**Comment by milanglacier** [comment]
> With rose-pine I can, but the kickstarter configuration is using catppuccin and that works (see screenshots). If catppuccin doesn't work for you out of the box, please try it with the kickstarter configuration so that we can compare a common ground.

Yes I made a mistake here. `Catppuccin` provides proper highlight for basic markups: bold, and italics.

**Comment by milanglacier** [comment]
> And I'll always prefer the more general solution, both for elegance and because quarto is meant to work with different languages, not just R.

Yes I configured `iron.nvim` for working with python (and of course working with R). And iron.nvim doesn't have problem with it. However `nvim-R` has better integration with R. (dynamic completions (like columns names of data frame)), an object browser (this is something different from lsp's outline). So I am still using Nvim-R for R.

One thing I am interested about: can slime send text from one buffer to different repls? I want to send code within a quarto file to multiple consoles like Python and R all together.

This is something not easy to complete with iron.nvim.

**Comment by milanglacier** [comment]
> That sounds like an nvimR error.
> 
> I honestly think the vim-slime way is better because there are less moving parts, but feel free to show your workflow! :) You can see my different workflows with it here: https://youtube.com/playlist?list=PLabWm-zCaD1axcMGvf7wFxJz8FZmyHSJ7

Here's the problem. If Nvim-R works solely fine, and only has problem when treesitter's markdown parser is installed, and pandoc-syntax is used, does this count as Nvim-R's problem?

**Comment by jmbuhr** [comment]
Well, yes. We are not throwing the error.

nvimR is solving some of the same problems in a more specialised way but there is no way for the general solution to anticipate edgecases thrown in by specialized tools. Otherwise it would become a specialized solution itself and break other things.

About automatically sending code to different REPLs: I go over this in video 3. tldr. I haven't built this, but it should be straightforward to detect the language context from TS (I am doing something like this for the code completion in otter.nvim based on a helper function from nvim-cmp) and change the current slime target based on that. If you end up making this I will be very happy about a PR to the quarto-nvim-kickstarter!

**Comment by milanglacier** [comment]
> About automatically sending code to different REPLs: I go over this in video 3. tldr. I haven't built this, but it should be straightforward to detect the language context from TS (I am doing something like this for the code completion in otter.nvim based on a helper function from nvim-cmp) and change the current slime target based on that. If you end up making this I will be very happy about a PR to the quarto-nvim-kickstarter!

My update on `iron.nvim`:

`iron.nvim` has the support of attaching one buffer to multiple REPLs by using `Iron attach r`, `Iron attach python`

So I think I am facing the similar problem as yours: using treesitter to detect the language context and fire an autocmd to automatically calls `Iron attach xxx` to switch the context.

**Comment by milanglacier** [comment]
> > That sounds like an nvimR error.
> > I honestly think the vim-slime way is better because there are less moving parts, but feel free to show your workflow! :) You can see my different workflows with it here: https://youtube.com/playlist?list=PLabWm-zCaD1axcMGvf7wFxJz8FZmyHSJ7
> 
> Here's the problem. If Nvim-R works solely fine, and only has problem when treesitter's markdown parser is installed, and pandoc-syntax is used, does this count as Nvim-R's problem? Or if Nvim-R only has problem when working with quarto-nvim, does this count as Nvim-R's problem?

This issue is already fixed by Nvim-R.

**Comment by jmbuhr** [comment]
Alright, I'll still leave this issue open in case you find out more about rose-pine and to notify you when I end up creating a quarto grammar as an extension of the markdown grammar, got some good tips by the nvim-treesitter people at https://github.com/nvim-treesitter/nvim-treesitter/pull/4212

**Comment by ThSGM** [comment]
I'm chiming here just to follow this discussion but to also note that, like @milanglacier I'm having issues with very poor syntax highlighting...to the extent that I basically had to disable the plugin and go back to using quarto-vim.

**Comment by jmbuhr** [comment]
This is no either-or question, you can use the regex-based syntax highlighting for pandoc/quarto together with treesitter based highlighting. I am testing this here: https://github.com/jmbuhr/quarto-nvim-kickstarter/tree/quarto-ft

It is now almost ready to be merged, but I need to decide on the least intrusive defaults (e.g. the conceallevel).

**Comment by jmbuhr** [comment]
It has now been merged into https://github.com/jmbuhr/quarto-nvim-kickstarter

**Comment by jmbuhr** [comment]
I will dedicate the next episode of "A coffee with quarto" to ways of customizing highlighting and conceals.

**Comment by jmbuhr** [comment]
Quarto might get its own treesitter grammer that is more detailed than the basic markdown grammar some time this year, after which I will favor this over the regex based vim-syntax workaround. Until then I consider this issue closed. Let me know if you run into other, specific issues.

**Comment by rhaynes74** [comment]
Hi folks, what are the essential pieces to get the math highlighting to work if you are not using kickstarter?
